src/AutoRest.CSharp/Common/Output/Models/Types/ModelFactoryTypeProvider.cs

```diff
@@ -29,19 +32,28 @@ internal sealed class ModelFactoryTypeProvider : TypeProvider
         protected override string DefaultName { get; }
         protected override string DefaultAccessibility { get; }
 
-        private IEnumerable<Method>? _methods;
-        public IEnumerable<Method> Methods => _methods ??= Models.Select(CreateMethod);
+        // TODO: remove this intermediate state once we generate it before output types
+        private IReadOnlyList<Method>? _methods;
+        private IReadOnlyList<Method> ShouldNotBeUsedForOutPut([CallerMemberName] string caller = "")
+        {
+            Debug.Assert(caller == nameof(OutputMethods) || caller == nameof(SignatureType), $"This method should not be used for output. Caller: {caller}");
```

Comment:
is it really necessary to add this kind of guard on a private method?
If this is internal I would agree that we need a guard to avoid abuse of the method.

Reply:
sounds good

Reply:
Had a long discussion with @m-nash, we would like to avoid the misusage of this within this class even.

Reply:
OK - we might need to add some document on this method to explicitly document it.

Reply:
Added